### PR TITLE
Add retain method to ListStore

### PR DESCRIPTION
**Pull request by ranfdev**
Motivation:
People often endup doing
```rust
for item in list_store.iter()
  if item.unwrap().name() == 'Giorgio' {
    list_store.remove(&item)
  }
}
```
But that `unwrap` is going to panic, because we added checks to prevent that an iterator over a list is mutated during iteration.

`retain` is just a nice alternative to provide an easier to use way to remove items while iterating the list.
```rust
list_store.retain(|item| item.name() != 'Giorgio')
```

It will also remove consecutive items with a single `splice` call, so If the user removes every item, `retain` doesn't spam `items_changed` signals

### CODE REVIEW

File: gio/src/list_store.rs [hunk 1]
```diff
@@ -76,6 +76,45 @@ impl ListStore {
...
+    // rustdoc-stripper-ignore-next
+    /// Retains only the elements specified by the predicate.
+    /// This method operates in place, visiting each element exactly once in the original order,
+    /// and preserves the order of the retained elements.
+    /// Because the elements are visited exactly once in the original order,
+    /// external state may be used to decide which elements to keep.
+    ///
+    /// # Panics
+    /// Panics if the predicate closure mutates the list by removing or adding items.
+    pub fn retain(&self, f: impl Fn(&glib::Object) -> bool) {
```
Comment (sdroege) [inline gio/src/list_store.rs:88]:
```suggestion
    pub fn retain(&self, f: impl FnMut(&glib::Object) -> bool) {
```

because why not :)

File: gio/src/list_store.rs [hunk 2]
```diff
@@ -76,6 +76,45 @@ impl ListStore {
...
+        let mut consec_removed = 0;
+        let mut i = 0;
+        const ADDITIONS: &[glib::Object] = &[]; // To satisfy the type checker
+
+        let changed = Rc::new(Cell::new(false));
+        let changed_clone = changed.clone();
+        let signal_id = self.connect_items_changed(move |_list, _, _, _| changed_clone.set(true));
```
Comment (sdroege) [inline gio/src/list_store.rs:95]: Ideally you'd like to store that in something that would disconnect on `Drop`

File: gio/src/list_store.rs [hunk 3]
```diff
@@ -76,6 +76,45 @@ impl ListStore {
...
+            }
+            if !keep {
+                consec_removed += 1;
+            } else if consec_removed > 0 {
+                self.splice(i - consec_removed, consec_removed, ADDITIONS);
+                changed.set(false);
+                i -= consec_removed;
+                consec_removed = 0;
+            }
+            i += 1;
```
Comment (sdroege) [inline gio/src/list_store.rs:110]: I hope the test covers the case of `consec_removed > 0` and afterwards `i += 1` and then checks that the next iteration ends up at the correct item?

That part is a bit hard to follow but otherwise this seems all correct :)

Comment (ranfdev) [reply gio/src/list_store.rs:126]: Eh, it should be covered. If it weren't, `retain` would remove or keep an unwanted item, but the test case checks that.

Comment (sdroege) [reply gio/src/list_store.rs:126]: I guess we should add coverage reports to the CI. That was very useful for these kind of things in gstreamer :) If it's covered, all good.

**Review by sdroege — APPROVED** [review]
